File: machine/ts/machine/runtime/timing/state.ts
import { HZ_SCALE } from '../../../spec/bmsx/timing';
import { calcCyclesPerFrameScaled } from './index';
import type { MachineModelSpec } from '../../../spec/bmsx/model';
import {
	GX_GPU_PCRTC_RESET_ACTIVE_DISPLAY_HALF_LINES,
	GX_GPU_PCRTC_RESET_REFRESH_UFPS_SCALED,
	GX_GPU_PCRTC_RESET_TOTAL_HALF_LINES,
} from '../../devices/gx/gpu_pcrtc';

export class TimingState {
	public ufps: number;
	public frameDurationMs: number;
	public cpuCyclesPerMillisecond: number;
	public pcrtcRevision = 0;
	public pcrtcRunning = true;
	public ufpsScaled = GX_GPU_PCRTC_RESET_REFRESH_UFPS_SCALED;
	public cpuHz: number;
	public cycleBudgetPerFrame: number;
	public totalHalfLines = GX_GPU_PCRTC_RESET_TOTAL_HALF_LINES;
	public activeDisplayHalfLines = GX_GPU_PCRTC_RESET_ACTIVE_DISPLAY_HALF_LINES;
	public geoWorkUnitsPerSec: number;

	constructor(model: MachineModelSpec) {
		this.cpuHz = model.cpuFreqHz;
		this.cycleBudgetPerFrame = calcCyclesPerFrameScaled(
			model.cpuFreqHz,
			GX_GPU_PCRTC_RESET_REFRESH_UFPS_SCALED,
		);
		this.geoWorkUnitsPerSec = model.geoWorkUnitsPerSec;
		this.ufps = this.ufpsScaled / HZ_SCALE;
		this.frameDurationMs = 1000 / this.ufps;
		this.cpuCyclesPerMillisecond = model.cpuFreqHz / 1000;
	}
}

File: machine/cpp/machine/runtime/timing/state.cpp
#include "machine/runtime/timing/state.h"

#include "machine/devices/gx/gpu_pcrtc.h"
#include "spec/bmsx/model.h"
#include "spec/bmsx/timing.h"
#include "machine/runtime/timing/index.h"

namespace bmsx {

TimingState::TimingState(const MachineModelSpec& model)
	: ufpsScaled(GX_GPU_PCRTC_RESET_REFRESH_UFPS_SCALED)
	, ufps(static_cast<f64>(GX_GPU_PCRTC_RESET_REFRESH_UFPS_SCALED) / static_cast<f64>(HZ_SCALE))
	, frameDurationMs(1000.0 / ufps)
	, pcrtcRunning(true)
	, totalHalfLines(GX_GPU_PCRTC_RESET_TOTAL_HALF_LINES)
	, activeDisplayHalfLines(GX_GPU_PCRTC_RESET_ACTIVE_DISPLAY_HALF_LINES)
	, cpuHz(model.cpuFreqHz)
	, cpuCyclesPerMillisecond(static_cast<f64>(model.cpuFreqHz) / 1000.0)
	, cycleBudgetPerFrame(calcCyclesPerFrameScaled(
		model.cpuFreqHz,
		GX_GPU_PCRTC_RESET_REFRESH_UFPS_SCALED))
	, geoWorkUnitsPerSec(model.geoWorkUnitsPerSec) {
}

} // namespace bmsx
